src/classes/Contacts.cls

```diff
@@ -0,0 +1,732 @@
+    }
+
+    public void onBeforeInsert() {
+        synchronizeNamingExclusionsFieldsBeforeInsert();
+    }
+
+    public void onBeforeUpdate() {
+        synchronizeNamingExclusionsFieldsBeforeUpdate();
+    }
+
+    public void onAfterInsert() {
+        processHouseholdsAfterInsert();
+    }
+
+    private void processHouseholdsAfterInsert() {
+        List<Contact> householdInserts = new List<Contact>();
+        List<Contact> contactEvaluations = new List<Contact>();
+        for (Contact con : (List<Contact>) getRecords()) {
+            if (needsHousehold(con)) {
+                householdInserts.add(con);
+            }
+            //If the user has chosen for only Contacts connected to Individual Accounts to have Households
+            //add them all to a list for evaluation
+            else if (needsEvaluation(con)) {
+                contactEvaluations.add(con);
+            } else if (needsHouseholdNameUpdated(con)) {
+                householdsNeedingNamesUpdated.add(householdIdFor(con));
+            }
+        }
+
+        if (householdsNeedingNamesUpdated.size() > 0) {
+            updateHouseholdNames(householdsNeedingNamesUpdated);
+        }
+
+        if (householdInserts.size() > 0) {
+            insertHouseholdsForContacts(householdInserts, true);
+        }
+
+        if (contactEvaluations.size() > 0) {
+            insertHouseholdsForContacts(
+                    getContactsNeedingHouseholds(contactEvaluations), true);
+        }
+    }
+
+    private Id householdIdFor(Contact con) {
+        return con.npo02__Household__c;
+    }
+
+    private Boolean needsHouseholdNameUpdated(Contact con) {
+        return householdIdFor(con) != null;
+    }
+
+    private Boolean needsEvaluation(Contact con) {
+        return householdIdFor(con) == null &&
+                settings.isAllIndividualsProcessor();
+    }
+
+    private Boolean needsHousehold(Contact con) {
+        return householdIdFor(con) == null
+                && settings.isAllProcessor();
+    }
+
+    private void buildOldHouseholdCollections() {
+        oldHouseholds.addAll(getOldHouseholdIds());
+        buildOldHouseholdSizeMap();
+    }
+
+    private void buildOldHouseholdSizeMap() {
+        for (AggregateResult aggregateResult : getMemberCountForOldHouseholds()) {
+            oldHouseholdSizeById.put(
+                    oldHouseholdIdFrom(aggregateResult),
+                    memberCountFrom(aggregateResult));
+        }
+
+        populateOldHouseholdSizeMapForEmptyHouseholds();
+    }
+
+    private void populateOldHouseholdSizeMapForEmptyHouseholds() {
+        for (Id hhid : oldHouseholds) {
+            if (oldHouseholdIsEmpty(hhid))
+                oldHouseholdSizeById.put(hhid, 0);
+        }
+    }
+
+    private Boolean oldHouseholdIsEmpty(Id hhid) {
+        return !oldHouseholdSizeById.containskey(hhid);
+    }
+
+    private Integer memberCountFrom(AggregateResult a) {
```

Comment:
Good idea!  I'll take a look.

Reply:
After a lot of back and forth, I moved this into the Households domain class and then replaced the Aggregate query.  It was being performed twice and had this manual augmentation to insert 0s for empty households since they didn't come back in the agg result.  I think the new change is much cleaner, which uses a subquery when the Households domain class is instantiated.  Let me know what you think!

see `Households.getHouseholdsById`